### PR TITLE
Modifications to "get_working_device" function in Pytorch. Added the capability to set a Pytorch device using the "set_device" function

### PR DESCRIPTION
**Pull request by lior-dikstein**
## Pull Request Description:
Refine PyTorch device management using DeviceManager singleton class.

- Adopted the Singleton pattern in `DeviceManager` for consistent and centralized management of PyTorch devices.
- Modified the `get_working_device` function in PyTorch for improved retrieval of the current device setting using the singleton `DeviceManager`.
- Implemented the `set_device` function using `DeviceManager`, enabling the capability to set the PyTorch device across various project parts dynamically.
- Incorporated flexibility in `DeviceManager` to handle specific CUDA device indices ('cuda:x') and the default CUDA device ('cuda').

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/__init__.py [hunk 1]
```diff
@@ -25,4 +25,4 @@
 from model_compression_toolkit.core.common.mixed_precision.kpi_tools.kpi import KPI
 from model_compression_toolkit.core.common.mixed_precision.mixed_precision_quantization_config import MixedPrecisionQuantizationConfig, MixedPrecisionQuantizationConfigV2
 from model_compression_toolkit.core.keras.kpi_data_facade import keras_kpi_data, keras_kpi_data_experimental
-from model_compression_toolkit.core.pytorch.kpi_data_facade import pytorch_kpi_data, pytorch_kpi_data_experimental
+from model_compression_toolkit.core.pytorch.kpi_data_facade import pytorch_kpi_data, pytorch_kpi_data_experimental
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/__init__.py:28]: Remove unneeded changes

File: model_compression_toolkit/core/pytorch/pytorch_device_config.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        return cls._instance
+
+    def set_device(self, device_name: str):
+        """
+        Set the device for PyTorch operations.
+
+        Args:
+            device_name (str): The name of the device, e.g., 'cuda:0' or 'cpu'.
+
+        If the specified device is not valid or available, it prints an error message without changing the current device.
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/pytorch_device_config.py:47]: Do you think we should alert or raise an error? If just a warning - fix the comment.

File: model_compression_toolkit/core/pytorch/pytorch_device_config.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        """
+        is_valid, message = self.is_valid_device(device_name)
+        if is_valid:
+            self.DEVICE = torch.device(device_name)
+        else:
+            logger.Logger.warning(message)
+
+    def get_device(self):
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/pytorch_device_config.py:55]: Add type hint

File: model_compression_toolkit/core/pytorch/pytorch_device_config.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        """
+        Get the current PyTorch device.
+
+        Returns:
+            torch.device: The current device set for PyTorch operations.
+        """
+        return self.DEVICE
+
+    @staticmethod
+    def is_valid_device(device_name: str):
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/pytorch_device_config.py:65]: returned type is missing

File: model_compression_toolkit/core/pytorch/pytorch_device_config.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                device_index = int(device_name.split(':')[1])
+                if device_index >= torch.cuda.device_count():
+                    return False, f"CUDA device index {device_index} out of range. Number of valid devices: {torch.cuda.device_count()}"
+            except IndexError:
+                # Handle cases where the device name is incorrectly formatted
+                return False, "Invalid CUDA device format. Use 'cuda' or 'cuda:x' where x is the device index."
+
+            return True, "Valid device"
+
+        return True, "Valid device"
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/pytorch_device_config.py:94]: What if it's 'cpu'? I think it should be checked here before True is returned

File: model_compression_toolkit/core/pytorch/pytorch_device_config.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+    Args:
+        device_name (str): The name of the device, e.g., 'cuda:0' or 'cpu'.
+
+    If the specified device is not valid or available, it prints an error message without changing the current device.
+    """
+    device_manager = DeviceManager()
+    device_manager.set_device(device_name)
+
+def get_working_device():
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/pytorch_device_config.py:110]: Type hints

**Review by reuvenperetz — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/core/pytorch/pytorch_device_config.py (hunk 2 above)
Comment (lior-dikstein) [reply model_compression_toolkit/core/pytorch/pytorch_device_config.py:47]: I think a warning. Changed the comment.

File: model_compression_toolkit/core/pytorch/pytorch_device_config.py (hunk 5 above)
Comment (lior-dikstein) [reply model_compression_toolkit/core/pytorch/pytorch_device_config.py:94]: There isn't a check for CPU because it is always a valid choice. This checks that if you choose a GPU it exists.

Comment (reuvenperetz) [reply model_compression_toolkit/core/pytorch/pytorch_device_config.py:96]: I think it's better to verify that it's 'cpu' and not unrelated string, but it's up to you

**Review by reuvenperetz — APPROVED** [review]
